src/peft/tuners/cartridge/utils.py

```diff
@@ -0,0 +1,215 @@
+    if config.peft_type != PeftType.CARTRIDGE:
+        raise ValueError(f"Adapter '{adapter_name}' is not a CARTRIDGE adapter (got {config.peft_type}).")
+    if num_virtual_tokens is None:
+        num_virtual_tokens = config.num_virtual_tokens
+
+    if use_chat_template and hasattr(tokenizer, "apply_chat_template"):
+        try:
+            input_ids = tokenizer.apply_chat_template(
+                [{"role": "system", "content": text}],
+                tokenize=True,
+                add_generation_prompt=False,
+                return_tensors="pt",
+            )
+        except TypeError:
+            # Some tokenizers don't support the full signature above.
+            input_ids = torch.tensor(tokenizer.apply_chat_template([{"role": "system", "content": text}]))[None, :]
+    else:
+        toks = tokenizer(text, return_tensors="pt", truncation=max_length is not None, max_length=max_length)
+        input_ids = toks["input_ids"]
+
+    input_ids = input_ids.to(model.device)
+    attention_mask = torch.ones_like(input_ids)
+    with model.disable_adapter():
+        outputs = model(input_ids=input_ids, attention_mask=attention_mask, use_cache=True)
+    return initialize_cartridge_from_past_key_values(
+        model,
+        adapter_name=adapter_name,
+        past_key_values=outputs.past_key_values,
+        num_virtual_tokens=num_virtual_tokens,
+    )
+
+
+def compose_cartridge_adapters(
+    adapter_paths: Sequence[str | Path],
+    *,
+    output_path: str | Path,
+    safe_serialization: bool = True,
+) -> None:
+    """
+    Compose multiple CARTRIDGE adapters by concatenating their prompt embeddings.
+
+    This implements the paper's "composition via concatenation" behavior at the adapter level (no runtime
+    multi-adapter).
+    """
+    adapter_paths = [Path(p) for p in adapter_paths]
+    if len(adapter_paths) < 2:
+        raise ValueError("Need at least 2 adapters to compose.")
+
+    configs = [PeftConfig.from_pretrained(str(p)) for p in adapter_paths]
+    for p, cfg in zip(adapter_paths, configs):
+        if cfg.peft_type != PeftType.CARTRIDGE:
+            raise ValueError(f"Adapter at '{p}' is not a CARTRIDGE adapter (got {cfg.peft_type}).")
+
+    base = configs[0]
+    for cfg in configs[1:]:
+        for attr in ("task_type", "token_dim", "num_layers", "num_attention_heads", "num_transformer_submodules"):
+            if getattr(cfg, attr, None) != getattr(base, attr, None):
+                raise ValueError(f"Incompatible CARTRIDGE configs for attribute '{attr}'.")
+
+    weights = [load_peft_weights(str(p), device="cpu") for p in adapter_paths]
+    prompt_embeddings = [w["prompt_embeddings"] for w in weights]
+    composed = torch.cat(prompt_embeddings, dim=0)
+
+    num_virtual_tokens = composed.shape[0]
+    # Preserve the "frozen prefix tokens" count of the first adapter only (matches a single attention-sink prefix).
+    num_frozen_tokens = int(getattr(base, "num_frozen_tokens", 0) or 0)
```

Comment:
This should be enough, as we already check that the config type is correct:

```suggestion
    num_frozen_tokens = base.num_frozen_tokens
```

Also, just wondering: Say we have one adapter with 1 frozen token and another with 2 frozen tokens, right now we would set it to 1 and ignore 2. I guess it's fine as this is intended for inference only, right?